src/daft-connect/src/lib.rs

```diff
@@ -0,0 +1,463 @@
+};
+use tonic::{transport::Server, Request, Response, Status};
+use tracing::{info, warn};
+use uuid::Uuid;
+
+use crate::{convert::map_to_tables, session::Session};
+
+mod command;
+mod config;
+mod convert;
+mod session;
+pub mod util;
+
+pub fn start(addr: &str) -> eyre::Result<()> {
+    info!("Daft-Connect server listening on {addr}");
+    let addr = util::parse_spark_connect_address(addr)?;
+
+    let service = DaftSparkConnectService::default();
+
+    info!("Daft-Connect server listening on {addr}");
+
+    std::thread::spawn(move || {
+        let runtime = tokio::runtime::Runtime::new().unwrap();
+        let result = runtime
+            .block_on(async {
+                Server::builder()
+                    .add_service(SparkConnectServiceServer::new(service))
+                    .serve(addr)
+                    .await
+            })
+            .wrap_err_with(|| format!("Failed to start server on {addr}"));
+
+        if let Err(e) = result {
+            eprintln!("Daft-Connect server error: {e:?}");
+        }
+
+        println!("done with runtime");
+
+        eyre::Result::<_>::Ok(())
+    });
+
+    Ok(())
+}
+
+#[derive(Default)]
+pub struct DaftSparkConnectService {
+    client_to_session: DashMap<Uuid, Session>, // To track session data
+}
+
+impl DaftSparkConnectService {
+    fn get_session(
+        &self,
+        session_id: &str,
+    ) -> Result<dashmap::mapref::one::RefMut<Uuid, Session>, Status> {
+        let Ok(uuid) = Uuid::parse_str(session_id) else {
+            return Err(Status::invalid_argument(
+                "Invalid session_id format, must be a UUID",
+            ));
+        };
+
+        let res = self
+            .client_to_session
+            .entry(uuid)
+            .or_insert_with(|| Session::new(session_id.to_string()));
+
+        Ok(res)
+    }
+}
+
+fn pretty_config() -> ron::ser::PrettyConfig {
+    ron::ser::PrettyConfig::default()
+        .extensions(
+            Extensions::IMPLICIT_SOME
+                | Extensions::UNWRAP_NEWTYPES
+                | Extensions::UNWRAP_VARIANT_NEWTYPES,
+        )
+        .indentor("  ".to_string())
+}
+
+#[tonic::async_trait]
+impl SparkConnectService for DaftSparkConnectService {
+    type ExecutePlanStream = std::pin::Pin<
+        Box<
+            dyn futures::Stream<Item = Result<ExecutePlanResponse, Status>> + Send + Sync + 'static,
+        >,
+    >;
+    type ReattachExecuteStream = std::pin::Pin<
+        Box<
+            dyn futures::Stream<Item = Result<ExecutePlanResponse, Status>> + Send + Sync + 'static,
+        >,
+    >;
+
+    #[tracing::instrument(skip_all)]
+    async fn execute_plan(
+        &self,
+        request: Request<ExecutePlanRequest>,
+    ) -> Result<Response<Self::ExecutePlanStream>, Status> {
+        let request = request.into_inner();
+
+        let plan_readable = ron::ser::to_string_pretty(&request, pretty_config()).unwrap();
+        info!("got plan\n{plan_readable}");
```

Comment:
is this just for debugging/logging? probably dont want to deserialize every request just for logging. should only deserialize and print if the user has `info` logging enabled.